src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs

```diff
@@ -0,0 +1,255 @@
+            {
+                hmac.AppendData(ikm);
+                GetHashAndReset(hmac, prk);
+            }
+        }
+
+        /// <summary>
+        /// Performs the HKDF-Expand function
+        /// See section 2.3 of <a href="https://tools.ietf.org/html/rfc5869#section-2.3">RFC5869</a>
+        /// </summary>
+        /// <param name="hashAlgorithmName">Hash algorithm used for HMAC operations</param>
+        /// <param name="prk">Pseudorandom key of at least <see cref="HashLength"/> bytes (usually the output from Expand step)</param>
+        /// <param name="outputLength">Length of the output keying material</param>
+        /// <param name="info">Optional context and application specific information</param>
+        /// <returns>Output keying material</returns>
+        public static byte[] Expand(HashAlgorithmName hashAlgorithmName, byte[] prk, int outputLength, byte[] info = null)
+        {
+            if (prk == null)
+                throw new ArgumentNullException(nameof(prk));
+
+            int hashLength = HashLength(hashAlgorithmName);
+
+            if (outputLength <= 0 || outputLength > 255 * hashLength)
+                throw new ArgumentOutOfRangeException(nameof(outputLength));
+
+            info = info ?? Array.Empty<byte>();
+            byte[] result = new byte[outputLength];
+            Expand(hashAlgorithmName, hashLength, prk, result, info);
+
+            return result;
+        }
+
+        /// <summary>
+        /// Performs the HKDF-Expand function
+        /// See section 2.3 of <a href="https://tools.ietf.org/html/rfc5869#section-2.3">RFC5869</a>
+        /// </summary>
+        /// <param name="hashAlgorithmName">Hash algorithm used for HMAC operations</param>
+        /// <param name="prk">Pseudorandom key of at least <see cref="HashLength"/> bytes (usually the output from Expand step)</param>
+        /// <param name="output">Output buffer representing output keying material</param>
+        /// <param name="info">Context and application specific information (can be an empty span)</param>
+        public static void Expand(HashAlgorithmName hashAlgorithmName, ReadOnlySpan<byte> prk, Span<byte> output, ReadOnlySpan<byte> info)
+        {
+            int hashLength = HashLength(hashAlgorithmName);
+
+            // We want to throw different exception type for this overload
+            if (output.Length > 255 * hashLength)
+                throw new ArgumentException(nameof(output));
+
+            Expand(hashAlgorithmName, hashLength, prk, output, info);
+        }
+
+        private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength, ReadOnlySpan<byte> prk, Span<byte> output, ReadOnlySpan<byte> info)
+        {
+            if (prk.Length < hashLength)
+                throw new ArgumentException(nameof(prk));
+
+            if (output.Length > 255 * hashLength)
+                throw new ArgumentOutOfRangeException(nameof(output));
```

Comment:
Didn't both Expand calls already deal with this?

Reply:
There is also DeriveKey but will move the logic directly under APIs